### PR TITLE
[TfL] auto close reports

### PR DESCRIPTION
**Pull request by struan**
Fixes https://github.com/mysociety/fixmystreet-commercial/issues/1955

[skip changelog]

### CODE REVIEW

**Review by dracos — CHANGES_REQUESTED** [review]

File: perllib/FixMyStreet/Script/TfL/AutoClose.pm [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        category => { -in => [ keys %$categories ] },
+        'me.state' => 'action scheduled',
+        bodies_str => $self->body->id,
+        'comments.state' => 'confirmed',
+        'comments.problem_state' => 'action scheduled',
+    },
+    {
+        group_by => 'me.id',
+        join => [ 'comments' ],
+        having => \[ 'MIN(comments.confirmed) < ?', $dtf->format_datetime($self->newest) ]
```
[the remark this answers is not in the document]
Comment (struan) [reply perllib/FixMyStreet/Script/TfL/AutoClose.pm:80]: I had originally just used the lastupdate of the problem which is problematic in the other direction.

fetching all the the updates does seem like the only way to resolve this though. I guess with this query as the basis then at least you are only considering reports where there is a chance that they are valid so it's not too awful.

File: perllib/FixMyStreet/Script/TfL/AutoClose.pm [hunk 2]
```diff
@@ -84,6 +84,20 @@ sub close_reports {
 
     my $count = 0;
     for my $r ( $reports->all ) {
+        my $comments = FixMyStreet::DB->resultset('Comment')->search(
+            { problem_id => $r->id },
+            { order_by => 'confirmed' }
+        );
+        my $earliest;
+        while ( my $c = $comments->next ) {
+            warn $c->problem_state . "\n";
```
Comment (dracos) [inline perllib/FixMyStreet/Script/TfL/AutoClose.pm:93]: Debug

File: perllib/FixMyStreet/Script/TfL/AutoClose.pm [hunk 3]
```diff
@@ -84,6 +84,20 @@ sub close_reports {
...
+            if ( $c->problem_state ne 'action scheduled' ) {
+                $earliest = undef;
+                next;
+            }
+            $earliest = $c->confirmed unless defined $earliest;
+        }
+        next unless defined $earliest && $earliest < $self->newest;
```
Comment (dracos) [inline perllib/FixMyStreet/Script/TfL/AutoClose.pm:100]: newest could be a lazy has attribute, so that it's only calculated once?

**Review by dracos — CHANGES_REQUESTED** [review]
Something odd happening in the tests, now they're running - failing sometimes, I think the test is assuming the order it will get a result in or something like that. Other than that, only one debug and one suggestion :)

File: perllib/FixMyStreet/Script/TfL/AutoClose.pm [hunk 4]
```diff
@@ -36,6 +36,17 @@ sub close {
     $self->close_reports($categories);
 }
 
+has newest => (
+    is => 'lazy',
+    isa => Maybe[InstanceOf['DateTime']],
```
Comment (dracos) [inline perllib/FixMyStreet/Script/TfL/AutoClose.pm:41]: Why Maybe, isn't it always a DateTime? :)

**Review by dracos — APPROVED** [review]
Looks good; one query about why something that I think is always set is a Maybe is all :)

**Pull request merged**